## Exercise 14.1

In [4]:
import numpy as np
import scipy as sy


In [32]:
g = lambda x: 0.1*x ** 4 - 1.5 *x ** 3 + 0.53 * x ** 2 + 2 * x + 1

def integration(f, a, b, N, method):
    if method == "midpoint":
        xspace = np.array([a + ((2*i + 1) * (b - a)) / (2 *N) for i in range(N)])
        return ((b - a)/ N) * np.sum(g(xspace))
    elif method == "trapezoid":
        xspace = np.array([a + (i * (b - a)) / (N) for i in range(N + 1)])
        return ((b - a) / (2*N)) * (g(xspace[0]) + 2*np.sum(g(xspace[1:N-1])) + g(xspace[N]))
    elif method == "Simpsons":
        xspace = np.array([a + (i * (b - a)) / (2*N) for i in range(2*N + 1)])
        part1 = (b - a) / (3*(2*N + 1))
        firstsum = 0
        secondsum = 0
        for i in range(1, len(xspace) - 1):
            if i % 2 == 0:
                secondsum = g(xspace[i]) + secondsum
            else:
                firstsum = g(xspace[i]) + firstsum
        part2 = (g(xspace[0]) + 4*firstsum + 2*secondsum + g(xspace[2*N]))
        return part1 * part2
    else:
        return None

a = -10
b = 10
N = 100

print("Midpoint :", integration(g, a, b, N, 'midpoint'))
print("Trapezoid :", integration(g, a, b, N, 'trapezoid'))
print("Simpsons :", integration(g, a, b, N, 'Simpsons'))

Midpoint : 4371.96476
Trapezoid : 4459.654288
Simpsons : 4351.5754825870645


## Exercise 14.2

In [34]:
from scipy.stats import norm 

In [55]:
def newton_cotes_z(mu, sigma, N, k):
    
    Z = np.linspace(mu - k*sigma, mu + k*sigma, N)
    omega = np.zeros_like(Z)
    omega[0] = norm.cdf((Z[0] + Z[1]) / 2, loc=mu, scale=sigma)
    omega[N - 1] = 1 - norm.cdf((Z[N - 2] + Z[N - 1]) / 2, loc=mu, scale=sigma)
    for i in range(1, N - 2):
        z_min = (Z[i - 1] + Z[i]) / 2
        z_max = (Z[i] + Z[i + 1]) / 2
        omega[i] = norm.cdf(z_max, loc=mu, scale=sigma) - norm.cdf(z_min, loc=mu, scale=sigma)
    return Z, omega


In [56]:
mu = 0
k = 3
sigma= 1
N = 11
newton_cotes_z(mu, sigma, N, k)

(array([-3. , -2.4, -1.8, -1.2, -0.6,  0. ,  0.6,  1.2,  1.8,  2.4,  3. ]),
 array([0.00346697, 0.01439745, 0.04894278, 0.11725292, 0.19802845,
        0.23582284, 0.19802845, 0.11725292, 0.04894278, 0.        ,
        0.00346697]))

## Exercise 14.3

In [64]:
def lognormal_dist(f, mu, sigma, N, k):
    logA, omega = f(mu, sigma, N, k)
    A = np.exp(logA)
    return A, omega

In [65]:
mu = 0
k = 3
sigma= 1
N = 11

lognormal_dist(newton_cotes_z, mu, sigma, N, k)

(array([ 0.04978707,  0.09071795,  0.16529889,  0.30119421,  0.54881164,
         1.        ,  1.8221188 ,  3.32011692,  6.04964746, 11.02317638,
        20.08553692]),
 array([0.00346697, 0.01439745, 0.04894278, 0.11725292, 0.19802845,
        0.23582284, 0.19802845, 0.11725292, 0.04894278, 0.        ,
        0.00346697]))

## Exercise 14.4

In [78]:
mu = 10.5
sigma = 0.8
N = 100
k = 3
Y, omega = lognormal_dist(newton_cotes_z, mu, sigma, N, k)
Y @ omega

49739.23910845465

In [72]:
np.exp(mu + sigma ** 2 / 2)

50011.087008521754

## Exercise 14.5 ********************

In [10]:
g = lambda x: 0.1*x ** 4 - 1.5 *x ** 3 + 0.53 * x ** 2 + 2 * x + 1
a = -10
b = 10
N = 3

def approx_gaussian_quad(params):
    a = -10
    b = 10
    N = 100
    w1, w2, w3, x1, x2, x3 = params
    w = np.array([w1, w2, w3])
    x = np.array([x1, x2, x3])
    equations = np.empty(6)
    for n in range(1, 7):
        equations[n - 1] = (b**n - a**n) / n - w @(x ** (n-1))
    return equations

In [11]:
from scipy.optimize import root
import numpy as np
output = root(approx_gaussian_quad, np.ones(6))
w = output.x[:3]
x = output.x[3:]
np.sum(w*g(x))

4373.333333340381

## Exercise 14.6

In [81]:
sy.integrate.quad(g, a, b)[0]

4373.333333333334

The right number is 4373.3333...  So this is pretty accurate.

## Exercise 14.7

In [82]:
import random

In [5]:
def monte_carlo(g, omega, N):
    x = np.random.uniform(omega[0][0], omega[0][1], N)
    y = np.random.uniform(omega[1][0], omega[1][1], N)
    return (4 / N) * np.sum(g(x, y))


In [15]:
boundary = [[-1, 1], [-1, 1]]
g = lambda x, y: (x**2 + y ** 2 <= 1)
monte_carlo(g, boundary, 100000000)

3.14155868

It takes aprox. around 100_000_000 to get to 3.1415

## Exercise 14.8 *****

## Exercise 14.9 ****